File: 917.Reverse-Only-Letters.go
// https://leetcode.com/problems/reverse-only-letters/description/
//
// algorithms
// Easy (58.82%)
// Total Accepted:    5k
// Total Submissions: 8.5k
// beats 100.0% of golang submissions

package leetcode

func reverseOnlyLetters(S string) string {
	tail := len(S) - 1
	byteS := []byte(S)

outer:
	for idx, ch := range byteS {
		if idx >= tail {
			break
		}
		if (ch >= 97 && ch <= 122) || (ch >= 65 && ch <= 90) {
			for {
				if (byteS[tail] >= 97 && byteS[tail] <= 122) || (byteS[tail] >= 65 && byteS[tail] <= 90) {
					byteS[idx] = byteS[tail]
					byteS[tail] = ch
					tail--
					break
				}
				tail--
				if tail == idx {
					break outer
				}
			}
		}
	}

	return string(byteS)
}

File: 917.Reverse-Only-Letters.java
// https://leetcode.com/problems/reverse-only-letters/
//
// algorithms
// Easy (56.7%)
// Total Accepted: 40,649
// Total Submissions: 71,689


class Solution {
    public String reverseOnlyLetters(String S) {
        char[] chs = S.toCharArray();
        int len = chs.length;
        int i = 0, j = len - 1;

        while (i < j) {
            while (i < j && !isLetter(chs[i])) {
                i++;
            }
            while (i < j && !isLetter(chs[j])) {
                j--;
            }

            swap(chs, i, j);
            i++;
            j--;
        }

        return new String(chs);
    }

    private void swap(char[] chs, int i, int j) {
        char tmp = chs[i];
        chs[i] = chs[j];
        chs[j] = tmp;
    }

    private boolean isLetter(char ch) {
        int tmp = ch - 'a';
        if (tmp >= 0 && tmp <= 25) {
            return true;
        }

        tmp = ch - 'A';
        if (tmp >= 0 && tmp <= 25) {
            return true;
        }

        return false;
    }

}